dev/TeachingTip/TeachingTip.cpp

```diff
@@ -0,0 +1,1540 @@
+        }
+        else if (actionText.size() > 0)
+        {
+            if (IsLightDismissEnabled())
+            {
+                winrt::VisualStateManager::GoToState(*this, L"ActionButtonVisible"sv, false);
+            }
+            else
+            {
+                //Without light dismiss we require that atleast one close button be shown at all times.
+                winrt::VisualStateManager::GoToState(*this, L"BothButtonsVisible"sv, false);
+            }
+        }
+        else if (closeText.size() > 0)
+        {
+            winrt::VisualStateManager::GoToState(*this, L"CloseButtonVisible"sv, false);
+        }
+        else
+        {
+            if (IsLightDismissEnabled())
+            {
+                winrt::VisualStateManager::GoToState(*this, L"NoButtonsVisible"sv, false);
+            }
+            else
+            {
+                //We require that atleast one close button be shown at all times.
+                winrt::VisualStateManager::GoToState(*this, L"CloseButtonVisible"sv, false);
+            }
+        }
+        break;
+    }
+}
+
+void TeachingTip::UpdateDynamicBleedingContentPlacementToTop()
+{
+    if (BleedingImagePlacement() == winrt::TeachingTipBleedingImagePlacementMode::Auto)
+    {
+        winrt::VisualStateManager::GoToState(*this, L"BleedingContentTop"sv, false);
+        if (m_currentBleedingEffectivePlacementMode != winrt::TeachingTipBleedingImagePlacementMode::Top)
+        {
+            m_currentBleedingEffectivePlacementMode = winrt::TeachingTipBleedingImagePlacementMode::Top;
+            TeachingTipTestHooks::NotifyEffectiveBleedingPlacementChanged(*this);
+        }
+    }
+}
+
+void TeachingTip::UpdateDynamicBleedingContentPlacementToBottom()
+{
+    if (BleedingImagePlacement() == winrt::TeachingTipBleedingImagePlacementMode::Auto)
+    {
+        winrt::VisualStateManager::GoToState(*this, L"BleedingContentBottom"sv, false);
+        if (m_currentBleedingEffectivePlacementMode != winrt::TeachingTipBleedingImagePlacementMode::Bottom)
+        {
+            m_currentBleedingEffectivePlacementMode = winrt::TeachingTipBleedingImagePlacementMode::Bottom;
+            TeachingTipTestHooks::NotifyEffectiveBleedingPlacementChanged(*this);
+        }
+    }
+}
+
+void TeachingTip::OnIsOpenChanged()
+{
+    if (IsOpen())
+    {
+        m_lastCloseReason = winrt::TeachingTipCloseReason::Programmatic;
+        if (m_target)
+        {
+            m_currentTargetBounds = m_target.get().TransformToVisual(nullptr).TransformBounds({
+                0.0,
+                0.0,
+                static_cast<float>(m_target.get().as<winrt::FrameworkElement>().ActualWidth()),
+                static_cast<float>(m_target.get().as<winrt::FrameworkElement>().ActualHeight())
+                });
+        }
+        if (!m_popup)
+        {
+            m_popup.set(winrt::Popup());
+            m_popup.get().Child(*this);
+            m_popup.get().Closed({ this, &TeachingTip::OnPopupClosed });
```

Comment:
I'm concerned about lifetime of this -- I think we either need auto_revoke or change this to get_strong()/get_weak() as appropriate.